Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/RetryMechanism.kt

```diff
@@ -0,0 +1,77 @@
+package de.rki.coronawarnapp.util
+
+import timber.log.Timber
+import kotlin.math.pow
+import kotlin.math.roundToLong
+
+object RetryMechanism {
+
+    fun <T> retryWithBackOff(
+        delayCalculator: (Attempt) -> Long? = createDelayCalculator(),
+        delayOperation: (Long) -> Unit = { Thread.sleep(it) },
+        retryCondition: (Attempt) -> Boolean = { true },
+        action: () -> T
+    ): T {
+        var current = Attempt()
+        while (true) {
+            Timber.v("Executing attempt: %s", current)
+            try {
+                return action()
+            } catch (e: Exception) {
+                current = current.copy(exception = e)
+            }
+
+            if (!retryCondition(current)) throw current.exception!!
+
+            val newDelay = delayCalculator(current)
+
+            if (newDelay == null) {
+                Timber.w("Retrycondition exceeded: %s", current)
+                throw current.exception!!
+            } else {
+                delayOperation(newDelay)
+            }
+
+            current = current.copy(
+                count = current.count + 1,
+                lastDelay = newDelay,
+                totalDelay = current.totalDelay + newDelay
+            )
+        }
+    }
+
+    private const val DEFAULT_TOTAL_MAX_RETRY = 15 * 1000L // 15 seconds total delay
```

Comment:
Minor, but shouldn't constants be on top of the class

Reply:
Good question. For classes they should be in the companion object which are supposed to be at the end of the class according to KotlinStyleGuide. Here this is an `object` already. I placed it where it is so it is close to where it is used. I would go with your suggestion if it were used in other locations too. Here it seems beneficial to view it all without scrolling, what do you think?